### PR TITLE
Fix simplified layer norm fp16 overflow

### LINKED ISSUE #29034: [CUDA] SkipSimplifiedLayerNormalization fp16 silently returns all zeros for large-magnitude inputs (variance accumulated in fp16)
### Describe the issue

The CUDA `SkipSimplifiedLayerNormalization` kernel (fp16, default non-strict mode) silently returns **all-zero outputs** when the input row magnitude is large. The default kernel accumulates the variance in the activation dtype `T`: for fp16, `rldval * val` ≈ `val² / hidden_size` overflows fp16's max (65504) once `|val|` exceeds roughly `sqrt(65504 * hidden_size)` per element, the sum of squares becomes `inf`, and the normalized output collapses to exact zeros — with no error, warning, or NaN to signal the failure.

Relevant code (v1.24.2):
- Non-strict kernels accumulate in `T`: [`skip_layer_norm_impl.cu#L93-L130`](https://github.com/microsoft/onnxruntime/blob/v1.24.2/onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu#L93-L130) (`cub::KeyValuePair<T, T>(rldval, rldval * val)`) and the vectorized variant ([`#L131-L180`](https://github.com/microsoft/onnxruntime/blob/v1.24.2/onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu#L131-L180), `T rldvalsq_sum`).
- Strict mode dispatches to `HostApplyLayerNorm<CudaT, float, CudaT, Simplified>` ([`skip_layer_norm.cc#L92-L109`](https://github.com/microsoft/onnxruntime/blob/v1.24.2/onnxruntime/contrib_ops/cuda/bert/skip_layer_norm.cc#L92-L109)), which accumulates in fp32 and produces correct results.
- The CPU implementation also accumulates in fp32, so CPU vs CUDA comparison shows the divergence directly.

**Real-world impact:** this is not a synthetic edge case. The fp16 ONNX export of [`openai/privacy-filter`](https://huggingface.co/openai/privacy-filter) (a gpt-oss-architecture PII NER model, `onnx/model_fp16.onnx`) legitimately produces hidden states with absmax ≈ 7,600 by its final layer. On CUDA the final `SkipSimplifiedLayerNormalization` therefore zeroes the hidden states and **every logit comes back exactly 0.0** — the privacy filter silently stops detecting any PII while appearing to run fine. The same model on `CPUExecutionProvider` works correctly, and on CUDA with `enable_skip_layer_norm_strict_mode=1` reaches 100% argmax agreement with CPU.

Observed on ONNX Runtime GPU 1.24.2 and 1.23.2 (Tesla T4, sm75). The analysis suggests it is not architecture-specific, since the accumulation dtype does not depend on SM version.

Suggested resolution: accumulate the mean/variance reduction in fp32 in the non-strict fp16/bf16 kernels as well (matching the CPU EP and strict mode); alternatively default `Simplified` fp16 to the strict path, or document the overflow hazard and the strict-mode workaround on the operator/provider-options pages. Happy to provide more details.

### To reproduce

```python
# pip install onnxruntime-gpu numpy onnx
import numpy as np
import onnx
from onnx import TensorProto, helper
import onnxruntime as ort

HIDDEN = 640

gamma = np.ones(HIDDEN, np.float16)
node = helper.make_node(
    "SkipSimplifiedLayerNormalization", ["X", "S", "G"], ["Y"],
    domain="com.microsoft", epsilon=1e-5,
)
graph = helper.make_graph(
    [node], "sln",
    [helper.make_tensor_value_info("X", TensorProto.FLOAT16, ["B", "T", HIDDEN]),
     helper.make_tensor_value_info("S", TensorProto.FLOAT16, ["B", "T", HIDDEN])],
    [helper.make_tensor_value_info("Y", TensorProto.FLOAT16, ["B", "T", HIDDEN])],
    [helper.make_tensor("G", TensorProto.FLOAT16, [HIDDEN], gamma.tobytes(), raw=True)],
)
model = helper.make_model(graph, opset_imports=[
    helper.make_opsetid("", 17), helper.make_opsetid("com.microsoft", 1)])

so = ort.SessionOptions()
so.log_severity_level = 3
cpu = ort.InferenceSession(model.SerializeToString(), so, providers=["CPUExecutionProvider"])
cuda = ort.InferenceSession(model.SerializeToString(), so,
                            providers=[("CUDAExecutionProvider", {"device_id": 0})])
cuda_strict = ort.InferenceSession(model.SerializeToString(), so,
                                   providers=[("CUDAExecutionProvider",
                                               {"device_id": 0,
                                                "enable_skip_layer_norm_strict_mode": "1"})])

print(f"onnxruntime {ort.__version__}")
rng = np.random.default_rng(0)
for scale in (1, 100, 500, 1000, 4000):
    x = (rng.standard_normal((1, 8, HIDDEN)) * scale).astype(np.float16)
    s = np.zeros_like(x)
    feeds = {"X": x, "S": s}
    yc = cpu.run(None, feeds)[0].astype(np.float32)
    yg = cuda.run(None, feeds)[0].astype(np.float32)
    ys = cuda_strict.run(None, feeds)[0].astype(np.float32)
    print(f"input_absmax={np.abs(x).max():8.1f}  cpu_absmax={np.abs(yc).max():6.3f}  "
          f"cuda_absmax={np.abs(yg).max():6.3f} (zeros={float((yg == 0).mean()):.2f})  "
          f"cuda_strict_absmax={np.abs(ys).max():6.3f}")
```

Output:

```text
onnxruntime 1.24.2
input_absmax=     3.9  cpu_absmax= 3.898  cuda_absmax= 3.900 (zeros=0.00)  cuda_strict_absmax= 3.898
input_absmax=   348.2  cpu_absmax= 3.504  cuda_absmax= 3.504 (zeros=0.00)  cuda_strict_absmax= 3.504
input_absmax=  2012.0  cpu_absmax= 4.008  cuda_absmax= 0.000 (zeros=1.00)  cuda_strict_absmax= 4.008
input_absmax=  3822.0  cpu_absmax= 3.791  cuda_absmax= 0.000 (zeros=1.00)  cuda_strict_absmax= 3.791
input_absmax= 17984.0  cpu_absmax= 4.395  cuda_absmax= 0.000 (zeros=1.00)  cuda_strict_absmax= 4.395
```

Expected: CUDA matches CPU (as it does with `enable_skip_layer_norm_strict_mode=1`).

### Urgency

Silent wrong results (all-zero logits) for fp16 transformer models with large hidden-state magnitudes — in our case it silently disabled a PII-redaction model in a privacy-sensitive pipeline. No crash or warning, so easy to ship unnoticed.

### Platform

Linux

### OS Version

Ubuntu 24.04 (6.17.0-1018-azure)

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.24.2 (also reproduced on 1.23.2)

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 12.8, cuDNN 9 (pip `nvidia-cudnn-cu12`), driver 610.43.02, Tesla T4 (sm75)

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
### Description

This PR resolves a critical numerical stability bug in the `SkipSimplifiedLayerNormalization` CUDA kernel when operating under `float16` (FP16) precision.

To prevent variance overflow during FP16 reduction, this change upgrades the precision of the intermediate reduction variables to `float` (FP32) within the non-strict simplified layer normalization paths:
1. **Precision Upgrade in Kernel Helpers (`layer_norm.cuh`)**: Modified the simplified variance reduction loops to accumulate intermediate squared states using `float` instead of `half` (via `cub::BlockReduce<float, TPB>`).
2. **Precision Upgrade in Host/Device Implementations (`skip_layer_norm_impl.cu`)**: Updated the shared memory allocation and thread-local data registers to use `float` for tracking the sum-of-squares.

This maintains FP16 inputs/outputs and retains high performance while guaranteeing numerical stability during reduction, achieving parity with the CPU EP and the strict-mode CUDA kernels.

### Motivation and Context

**Fixes #29034**

When executing transformer architectures in FP16 mode, input activations with large magnitudes (e.g., maximum absolute values > 7600, common in deep layer activations, scaling factors, or specific outlier features) can cause the variance calculation inside `SkipSimplifiedLayerNormalization` to overflow silently.

Because the maximum representable value for FP16 is $65504$, accumulating squared activations for a hidden dimension (e.g., $hidden\_size = 768$ or $1024$) easily exceeds this limit. In the existing non-strict kernels, this results in an overflow to infinity ($\infty$), producing `NaN` or silent all-zero outputs without throwing a runtime exception or crash. This is particularly problematic in production (e.g., `openai/privacy-filter` deployments), where the PII detection can silently fail by outputting zero logits.

By upgrading the intermediate accumulation to FP32, the numerical bounds are dramatically increased, fully eliminating the overflow risk for these models.

### CODE REVIEW

**Comment by SakshamKapoor2911** [comment]
@microsoft-github-policy-service agree
